File: windows/src/performance/performance_collector.cpp
#include "performance_collector.h"

#include <algorithm>

namespace mango_player {

PerformanceCollector::PerformanceCollector() {
    last_fps_calc_time_ = std::chrono::steady_clock::now();
}

PerformanceCollector::~PerformanceCollector() {
    Stop();
}

void PerformanceCollector::Configure(int sample_interval_ms,
                                      bool collect_bandwidth,
                                      bool collect_bitrate) {
    sample_interval_ms_ = sample_interval_ms;
    collect_bandwidth_ = collect_bandwidth;
    collect_bitrate_ = collect_bitrate;
}

void PerformanceCollector::SetCallback(std::shared_ptr<PerformanceCallback> callback) {
    callback_ = callback;
}

void PerformanceCollector::Start() {
    if (is_collecting_.exchange(true)) {
        return;  // Already collecting
    }
    
    last_fps_calc_time_ = std::chrono::steady_clock::now();
    frame_count_ = 0;
    
    collection_thread_ = std::thread(&PerformanceCollector::CollectionLoop, this);
}

void PerformanceCollector::Stop() {
    is_collecting_ = false;
    
    if (collection_thread_.joinable()) {
        collection_thread_.join();
    }
}

void PerformanceCollector::OnFrameRendered() {
    frame_count_++;
}

void PerformanceCollector::OnFrameDropped() {
    dropped_frames_++;
}

void PerformanceCollector::OnFrameDecoded(double video_time_ms, double audio_time_ms) {
    decoded_frames_++;
    video_decode_time_ms_ = video_time_ms;
    audio_decode_time_ms_ = audio_time_ms;
}

void PerformanceCollector::OnRenderComplete(double time_ms) {
    render_time_ms_ = time_ms;
}

void PerformanceCollector::SetVideoSize(int width, int height) {
    width_ = width;
    height_ = height;
}

void PerformanceCollector::SetHardwareDecoding(bool enabled) {
    is_hardware_decoding_ = enabled;
}

void PerformanceCollector::SetBufferLength(int buffer_ms) {
    buffer_length_ms_ = buffer_ms;
}

void PerformanceCollector::SetBitrate(int video_bps, int audio_bps) {
    video_bitrate_ = video_bps;
    audio_bitrate_ = audio_bps;
}

void PerformanceCollector::SetBandwidth(int bandwidth_bps) {
    bandwidth_bps_ = bandwidth_bps;
}

PerformanceMetrics PerformanceCollector::GetMetrics() {
    std::lock_guard<std::mutex> lock(metrics_mutex_);
    
    // 计算帧率
    auto now = std::chrono::steady_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
        now - last_fps_calc_time_).count();
    
    if (elapsed > 0) {
        current_fps_ = static_cast<double>(frame_count_) * 1000.0 / elapsed;
    }
    
    PerformanceMetrics metrics;
    metrics.frame_rate = current_fps_;
    metrics.dropped_frames = static_cast<int>(dropped_frames_.load());
    metrics.decoded_frames = static_cast<int>(decoded_frames_.load());
    metrics.video_decode_time_ms = video_decode_time_ms_;
    metrics.audio_decode_time_ms = audio_decode_time_ms_;
    metrics.render_time_ms = render_time_ms_;
    metrics.buffer_length_ms = buffer_length_ms_;
    metrics.is_hardware_decoding = is_hardware_decoding_;
    metrics.width = width_;
    metrics.height = height_;
    
    if (collect_bitrate_) {
        metrics.video_bitrate = video_bitrate_;
        metrics.audio_bitrate = audio_bitrate_;
    }
    
    if (collect_bandwidth_) {
        metrics.bandwidth_bps = bandwidth_bps_;
    }
    
    return metrics;
}

std::map<std::string, std::variant<double, int, bool>> PerformanceCollector::GetMetricsMap() {
    return GetMetrics().ToMap();
}

void PerformanceCollector::CollectionLoop() {
    while (is_collecting_) {
        std::this_thread::sleep_for(std::chrono::milliseconds(sample_interval_ms_));
        
        if (!is_collecting_) break;
        
        auto metrics = GetMetrics();
        
        // 重置帧计数
        frame_count_ = 0;
        last_fps_calc_time_ = std::chrono::steady_clock::now();
        
        // 调用回调
        if (auto callback = callback_.lock()) {
            callback->OnPerformanceUpdate(metrics);
        }
    }
}

}  // namespace mango_player

File: macos/Classes/Performance/PerformanceCollector.swift
import Foundation
import CoreMedia
import AVFoundation

/// macOS 性能数据收集器
///
/// 收集 FFmpeg/AVFoundation 播放过程中的性能指标，包括：
/// - 帧率 (FPS)
/// - 丢帧数
/// - 解码耗时
/// - 缓冲区状态
public class PerformanceCollector {
    
    // MARK: - Properties
    
    private var timer: Timer?
    private var isCollecting = false
    
    // 帧率计算
    private var frameCount: Int64 = 0
    private var lastFpsCalcTime: CFAbsoluteTime = 0
    private var currentFps: Double = 0
    private let frameCountLock = NSLock()
    
    // 丢帧统计
    private var droppedFrames: Int64 = 0
    private var decodedFrames: Int64 = 0
    
    // 解码耗时
    private var videoDecodeTimeMs: Double = 0
    private var audioDecodeTimeMs: Double = 0
    private var renderTimeMs: Double = 0
    
    // 视频信息
    private var width: Int = 0
    private var height: Int = 0
    private var isHardwareDecoding: Bool = false
    
    // 缓冲和码率
    private var bufferLengthMs: Int = 0
    private var videoBitrate: Int = 0
    private var audioBitrate: Int = 0
    private var bandwidthBps: Int = 0
    
    // 采样配置
    private var sampleIntervalMs: Int = 1000
    private var collectBandwidth = true
    private var collectBitrate = true
    
    // 回调
    public var callback: PerformanceCallback?
    
    // MARK: - Initialization
    
    public init() {}
    
    // MARK: - Public Methods
    
    /// 配置采样参数
    public func configure(
        sampleIntervalMs: Int = 1000,
        collectBandwidth: Bool = true,
        collectBitrate: Bool = true
    ) {
        self.sampleIntervalMs = sampleIntervalMs
        self.collectBandwidth = collectBandwidth
        self.collectBitrate = collectBitrate
    }
    
    /// 开始收集
    public func start() {
        guard !isCollecting else { return }
        isCollecting = true
        
        lastFpsCalcTime = CFAbsoluteTimeGetCurrent()
        frameCount = 0
        
        scheduleCollection()
    }
    
    /// 停止收集
    public func stop() {
        isCollecting = false
        timer?.invalidate()
        timer = nil
    }
    
    /// 记录一帧
    public func onFrameRendered() {
        frameCountLock.lock()
        frameCount += 1
        frameCountLock.unlock()
    }
    
    /// 记录丢帧
    public func onFrameDropped() {
        droppedFrames += 1
    }
    
    /// 记录解码完成
    public func onFrameDecoded(videoTimeMs: Double, audioTimeMs: Double) {
        decodedFrames += 1
        videoDecodeTimeMs = videoTimeMs
        audioDecodeTimeMs = audioTimeMs
    }
    
    /// 记录渲染耗时
    public func onRenderComplete(timeMs: Double) {
        renderTimeMs = timeMs
    }
    
    /// 设置视频尺寸
    public func setVideoSize(width: Int, height: Int) {
        self.width = width
        self.height = height
    }
    
    /// 设置硬件解码状态
    public func setHardwareDecoding(_ enabled: Bool) {
        isHardwareDecoding = enabled
    }
    
    /// 设置缓冲区长度
    public func setBufferLength(_ bufferMs: Int) {
        bufferLengthMs = bufferMs
    }
    
    /// 设置码率
    public func setBitrate(videoBps: Int, audioBps: Int) {
        videoBitrate = videoBps
        audioBitrate = audioBps
    }
    
    /// 设置带宽
    public func setBandwidth(_ bandwidthBps: Int) {
        self.bandwidthBps = bandwidthBps
    }
    
    /// 获取当前性能指标
    public func getMetrics() -> PerformanceMetrics {
        // 计算帧率
        let now = CFAbsoluteTimeGetCurrent()
        let elapsed = now - lastFpsCalcTime
        
        frameCountLock.lock()
        let currentFrameCount = frameCount
        frameCountLock.unlock()
        
        if elapsed > 0 {
            currentFps = Double(currentFrameCount) / elapsed
        }
        
        return PerformanceMetrics(
            frameRate: currentFps,
            droppedFrames: Int(droppedFrames),
            decodedFrames: Int(decodedFrames),
            videoDecodeTimeMs: videoDecodeTimeMs,
            audioDecodeTimeMs: audioDecodeTimeMs,
            renderTimeMs: renderTimeMs,
            bufferLengthMs: bufferLengthMs,
            bandwidthBps: collectBandwidth ? bandwidthBps : nil,
            isHardwareDecoding: isHardwareDecoding,
            videoBitrate: collectBitrate ? videoBitrate : nil,
            audioBitrate: collectBitrate ? audioBitrate : nil,
            width: width,
            height: height
        )
    }
    
    /// 转换为 Dictionary（用于 Platform Channel）
    public func getMetricsMap() -> [String: Any?] {
        return getMetrics().toMap()
    }
    
    // MARK: - Private Methods
    
    private func scheduleCollection() {
        guard isCollecting else { return }
        
        let interval = TimeInterval(sampleIntervalMs) / 1000.0
        timer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            guard let self = self, self.isCollecting else { return }
            
            let metrics = self.getMetrics()
            self.callback?.onPerformanceUpdate(metrics: metrics)
            
            // 重置帧计数
            self.frameCountLock.lock()
            self.frameCount = 0
            self.frameCountLock.unlock()
            self.lastFpsCalcTime = CFAbsoluteTimeGetCurrent()
        }
    }
}

// MARK: - PerformanceCallback Protocol

public protocol PerformanceCallback: AnyObject {
    func onPerformanceUpdate(metrics: PerformanceMetrics)
}

// MARK: - PerformanceMetrics

public struct PerformanceMetrics {
    public let frameRate: Double
    public let droppedFrames: Int
    public let decodedFrames: Int
    public let videoDecodeTimeMs: Double
    public let audioDecodeTimeMs: Double
    public let renderTimeMs: Double
    public let bufferLengthMs: Int
    public let bandwidthBps: Int?
    public let isHardwareDecoding: Bool
    public let videoBitrate: Int?
    public let audioBitrate: Int?
    public let width: Int?
    public let height: Int?
    
    public init(
        frameRate: Double,
        droppedFrames: Int,
        decodedFrames: Int,
        videoDecodeTimeMs: Double,
        audioDecodeTimeMs: Double,
        renderTimeMs: Double,
        bufferLengthMs: Int,
        bandwidthBps: Int?,
        isHardwareDecoding: Bool,
        videoBitrate: Int?,
        audioBitrate: Int?,
        width: Int?,
        height: Int?
    ) {
        self.frameRate = frameRate
        self.droppedFrames = droppedFrames
        self.decodedFrames = decodedFrames
        self.videoDecodeTimeMs = videoDecodeTimeMs
        self.audioDecodeTimeMs = audioDecodeTimeMs
        self.renderTimeMs = renderTimeMs
        self.bufferLengthMs = bufferLengthMs
        self.bandwidthBps = bandwidthBps
        self.isHardwareDecoding = isHardwareDecoding
        self.videoBitrate = videoBitrate
        self.audioBitrate = audioBitrate
        self.width = width
        self.height = height
    }
    
    public func toMap() -> [String: Any?] {
        return [
            "frameRate": frameRate,
            "droppedFrames": droppedFrames,
            "decodedFrames": decodedFrames,
            "videoDecodeTimeMs": videoDecodeTimeMs,
            "audioDecodeTimeMs": audioDecodeTimeMs,
            "renderTimeMs": renderTimeMs,
            "bufferLengthMs": bufferLengthMs,
            "bandwidthBps": bandwidthBps,
            "isHardwareDecoding": isHardwareDecoding,
            "videoBitrate": videoBitrate,
            "audioBitrate": audioBitrate,
            "width": width,
            "height": height
        ]
    }
    
    /// 性能等级
    public var level: PerformanceLevel {
        let dropRate = decodedFrames > 0 ? Double(droppedFrames) / Double(decodedFrames + droppedFrames) : 0
        if dropRate >= 0.05 {
            return .critical
        } else if dropRate >= 0.01 {
            return .warning
        } else {
            return .healthy
        }
    }
}

// MARK: - PerformanceLevel

public enum PerformanceLevel {
    case healthy   // 丢帧率 < 1%
    case warning   // 丢帧率 1% - 5%
    case critical  // 丢帧率 >= 5%
}
